Re-enable analyzers in the master branch

mavasani opened the pull request:
The [change](https://github.com/dotnet/roslyn/commit/6be56fadb4c417c4b6af0bf62f1242e41c8dd124) to move analyzers to 2.5 version broke analyzers in the branch - they are laid out differently in the new nuget package. Build targets are unable to find the props for importing the analyzers and silently bail out executing them. This PR fixes the following:

1. Fix the props file path in Imports.Targets
2. Move to a newer analyzer package which has an important [bug fix](https://github.com/dotnet/roslyn-analyzers/pull/1255)
3. Report a build error if analyzer props file does not exist
4. Fix the public API errors (mostly due to duplicate entries in shipped and unshipped files).
5. Fix one instance of RS0014 (Do not use Enumerable methods on indexable collections. Instead use the collection directly) introduced recently in the code.
6. Add suppression for one instance of CA1802 in WellKnownAttributeData.cs that fixed the signing issue.

heejaechang commented:
@mavasani 

/mnt/j/workspace/dotnet_roslyn/master/ubuntu_16_debug_prtest/build/Targets/Imports.targets(363,7): error : Analyzer packages have not been restored, run Restore.cmd before building [/mnt/j/workspace/dotnet_roslyn/master/ubuntu_16_debug_prtest/src/Compilers/CSharp/CscCore/CscCore.csproj]

are those only work with windows?

jasonmalinowski commented on build/Targets/Imports.targets:
```diff
@@ -360,6 +360,8 @@
   <Target Name="RestoreToolsetCheck" Condition="'$(BuildingProject)' == 'true'">
       <Error Text="Toolset packages have not been restored, run Restore.cmd before building"
              Condition="!Exists('$(ToolsetCompilerPropsFilePath)')" />
+      <Error Text="Analyzer packages have not been restored, run Restore.cmd before building"
```
Is this just a test that should really exist on the roslyn-analyzers side of things?

jasonmalinowski commented on build/Targets/Settings.props:
```diff
@@ -15,7 +15,7 @@
     <VSIXExpInstallerPackageVersion>0.2.1-beta</VSIXExpInstallerPackageVersion>
     <RoslynDiagnosticsNugetPackageVersion>$(MicrosoftNetRoslynDiagnosticsVersion)</RoslynDiagnosticsNugetPackageVersion>
-    <RoslynDiagnosticsPropsFilePath>$(NuGetPackageRoot)\Microsoft.Net.RoslynDiagnostics\$(RoslynDiagnosticsNugetPackageVersion)\roslyn\Microsoft.Net.RoslynDiagnostics.props</RoslynDiagnosticsPropsFilePath>
+    <RoslynDiagnosticsPropsFilePath>$(NuGetPackageRoot)\Microsoft.Net.RoslynDiagnostics\$(RoslynDiagnosticsNugetPackageVersion)\build\Microsoft.Net.RoslynDiagnostics.props</RoslynDiagnosticsPropsFilePath>
```
Should we just move to having a shared PackageReference here rather than doing this hack? Ditto for the compiler stuff.

mavasani replied:
Yes, I think we should do that change. Let me give that a try and if it turns out to be non-trivial, I'll file an issue to track that work.

mavasani commented:
@jasonmalinowski - addressed your feedback.

jasonmalinowski commented on src/Samples/VisualBasic/ConsoleClassifier/Program.vb:
```diff
@@ -21,10 +21,10 @@ Sub Main()
 End Module")
-        document = Await Formatter.FormatAsync(document)
-        Dim text As SourceText = Await document.GetTextAsync()
+        document = Await Formatter.FormatAsync(document).ConfigureAwait(False)
```
This shouldn't be running on samples though. We do it as a library so other apps don't have to think about it.

jaredpar replied:
Why don't we run on samples? My intuition, based on the setup of our build, is that it would run an samples.

mavasani commented:
Tagging @Srivatsn @jaredpar for project system help. For their context, this change moves Roslyn repo to use PackageReference for roslyn analyzers.

1. Regular windows PRs are running fine after this change - also validated analyzers are running as my second last commit failed with analyzer diagnostics.
2. `ubuntu_14_debug_prtest` and `ubuntu_16_debug_prtest` are failing with `error CS0006: Metadata file '/home/dotnet-bot/.nuget/packages/Microsoft.CodeQuality.Analyzers/2.5.0-beta1-61911-01/analyzers/dotnet/cs/Analyzer.Utilities.dll' could not be found [/mnt/j/workspace/dotnet_roslyn/master/ubuntu_14_debug_prtest/src/Compilers/Core/Portable/CodeAnalysis.csproj]`. Restore doesn't seem to work with package references on Ubuntu?
3. `windows_determinism_prtest` is failing with: `error AD0001: Analyzer 'Microsoft.NetCore.Analyzers.ImmutableCollections.DoNotCallToImmutableArrayOnAnImmutableArrayValueAnalyzer' threw an exception of type 'System.InvalidOperationException' with message 'Feature 'IOperation' is disabled.'.` Is Settings.props not getting imported for this job? See https://github.com/dotnet/roslyn/blob/master/build/Targets/Settings.props#L20

jaredpar commented on build/Targets/Imports.targets:
```diff
@@ -321,6 +321,14 @@
+
+  <ItemGroup Condition="'$(UseRoslynAnalyzers)' == 'true'">
+    <PackageReference Include="Microsoft.Net.RoslynDiagnostics" Version="$(RoslynDiagnosticsNugetPackageVersion)" />
+  </ItemGroup>
```
Can we make this a single item group and put the conditional on the PackageReference?

jaredpar commented:
> ubuntu_14_debug_prtest and ubuntu_16_debug_prtest are failing with error CS0006: Metadata file '/home/dotnet-bot/.nuget/packages/Microsoft.CodeQuality.Analyzers/2.5.0-beta1-61911-01/analyzers/dotnet/cs/Analyzer.Utilities.dll' could not be found [/mnt/j/workspace/dotnet_roslyn/master/ubuntu_14_debug_prtest/src/Compilers/Core/Portable/CodeAnalysis.csproj]. Restore doesn't seem to work with package references on Ubuntu?

Likely a path casing issue. NuGet restore will case the name of the package in all lower case. So it will be restored to `.nuget/packages/microsoft.codequality.analyzers/` not the casing listed in the error. 

> windows_determinism_prtest is failing with: error AD0001: Analyzer 'Microsoft.NetCore.Analyzers.ImmutableCollections.DoNotCallToImmutableArrayOnAnImmutableArrayValueAnalyzer' threw an exception of type 'System.InvalidOperationException' with message 'Feature 'IOperation' is disabled.'. Is Settings.props not getting imported for this job? See https://github.com/dotnet/roslyn/blob/master/build/Targets/Settings.props#L20

The determinism test passes `/p:Features="debug-determinism"` which is overriding the value in Settings.props. It also passes `/p:UseRoslynAnalyzers=false` hence that analyzer shoudln't be loading at all.

mavasani commented:
> Likely a path casing issue. NuGet restore will case the name of the package in all lower case. So it will be restored to .nuget/packages/microsoft.codequality.analyzers/ not the casing listed in the error.

@srivatsn Who owns this bug? I am going to revert back to non-PackageReference ItemGroup for non-Windows OSes.

jaredpar commented:
@mavasani, @srivatsn 

> Who owns this bug? I am going to revert back to non-PackageReference ItemGroup for non-Windows OSes.

Which bug? To be clear, the NuGet behavior is "By Design". Been that way for quite some time.

mavasani commented:
> Which bug? To be clear, the NuGet behavior is "By Design". Been that way for quite some time.

I am not sure how I could make this work. I am just adding a PackageReference to Microsoft.Net.RoslynDiagnostics, which has dependent NuGet packages, one of them being `Microsoft.CodeQuality.Analyzers`, for which the error is being reported.

srivatsn commented:
I suspect what's happening is that NuGet lays out the files on disk with lowercase but the assets file still has the other casing and hence the inputs passed to the compiler don't match the casing on disk. This must be specific to analyzers otherwise most packages likely would be broken on Linux.

dpoeschl commented:
retest windows_release_vs-integration_prtest please

mavasani commented:
> The determinism test passes /p:Features="debug-determinism" which is overriding the value in Settings.props. It also passes /p:UseRoslynAnalyzers=false hence that analyzer shoudln't be loading at all

The issue here seems to be that the restore step for determinism job [here](https://github.com/dotnet/roslyn/blob/master/build/scripts/test-determinism.ps1#L29) is causing analyzer package references to be included and added to project.assets.json. `/p:UseRoslynAnalyzers=false` flag is only passed on the build step subsequently, but that doesn't change the assets file and hence analyzers get executed. Do we know of a way to specify msbuild arguments when running nuget restore? Otherwise, for now I am going to just enable analyzer execution for determinism job and file a separate issue to fix this problem.

srivatsn commented:
@Emgarten is there a way to pass msbuild proeprties to nuget.exe? 

Other option is to use msbuild /t:restore instead of nuget.exe but that might be a larger change.

jaredpar commented:
> Other option is to use msbuild /t:restore instead of nuget.exe but that might be a larger change.

That won't work. The` /t:restore` target only works when VS is installed and that's not true in all of our cases. 

> The issue here seems to be that the restore step for determinism job here is causing analyzer package references to be included and added to project.assets.json. 

What changed here? This never happened before. Or is it that they've always been loaded and the new packages just have an issue when done under restore?

mavasani commented:
> What changed here? This never happened before. Or is it that they've always been loaded and the new packages just have an issue when done under restore?

Prior to this PR, we were not using PackageReferences for analyzer nuget package, instead we were doing an explicit import of the analyzer props file in Imports.targets, which in turn adds the analyzer references.

jaredpar commented:
How does PackageReference change the process though? Is it expected that analyzers should load during a restore? That doesn't seem right.

mavasani commented:
> How does PackageReference change the process though? Is it expected that analyzers should load during a restore? That doesn't seem right.

Analyzers are not loaded during restore, but the projects.assets.json file is written during this step. So if restore was executed with the default value of UseRoslynAnalyzers, which is true, then the analyzer package reference will be included in json file and hence get loaded during subsequent build step, regardless of the value of UseRoslynAnalyzers specified during build.

jaredpar commented:
Ah okay. I see the problem now.

mavasani commented:
@jaredpar Any concerns if I enabled analyzers for the determinism job and opened a separate tracking bug to fix this underlying issue and revert my workaround?

jaredpar commented:
> Any concerns if I enabled analyzers for the determinism job and opened a separate tracking bug to fix this underlying issue and revert my workaround?

Yeah we specifically didn't want analyzers running during that run. The idea is to do a compiler only test because that is all we actually guarantee determinism for.

mavasani commented:
Then I am not sure if there is any way to move analyzers to PackageReference unless we can specify msbuild arguments during NuGet restore. I am going to repurpose  #21005 to track the work of moving us to PackageReferences for toolset and analyzer packages and move back this PR to my first commit without any PackageReferences.

emgarten commented:
> @emgarten is there a way to pass msbuild proeprties to nuget.exe?

The env var `NUGET_RESTORE_MSBUILD_ARGS` will be passed from nuget.exe to msbuild during restore. 

Also, when restore runs it sets `ExcludeRestorePackageImports=true` which you *could* check with an  msbuild condition. This shouldn't be set any other time outside of restore.

Note, none of these are official extension points in nuget.

jaredpar commented:
@mavasani 

> Then I am not sure if there is any way to move analyzers to PackageReference unless we can specify msbuild arguments during NuGet restore

This still seems a bit broken from my perspective. Why do I need to re-restore in order to disable analyzers? The compiler team does that on a not uncommon basis to do investigations. That was a feature we've relied on for a long time now. Why are we breaking that functionality?

mavasani commented:
> This still seems a bit broken from my perspective. Why do I need to re-restore in order to disable analyzers? 

I agree, but our current design implicitly executes analyzers if you specify an analyzer reference and these are determined during the restore step in the PackageReference world. The only way to decouple this would be to add a feature to allow skipping analyzer execution by setting a new boolean property in the project.

mavasani commented:
FYI, I have moved back to the original approach of explicitly importing the analyzer props instead of using PackageReferences. So basically, its the original change on which I got signoffs + bunch of casing fixes for ubuntu + moving to latest analyzer and toolset compiler packages.

Moving to analyzer PackageReference is tracked by https://github.com/dotnet/roslyn/issues/21005, which needs to address Jared's concern: https://github.com/dotnet/roslyn/pull/20990#issuecomment-316863805.

mavasani commented:
@dotnet-bot retest windows_release_vs-integration_prtest please

jaredpar commented:
> I agree, but our current design implicitly executes analyzers if you specify an analyzer reference and these are determined during the restore step in the PackageReference world. 

That feels broken. Analyzers are an optional piece, they should have a way to be disabled. OTherwise it will break scenarios like this where disabling is advantageous.

jaredpar commented:
@mavasani we need to undo this PR. It is causing a major build time regression. On most machines it's trippling the time needed to build Rosyln.sln.

CyrusNajmabadi commented:
on mine it's a 6x regression.

mavasani commented:
Do we know what caused the regression? Does the regression go away if built with /p:UseRoslynAnalyzers = false? If not, is it a perfect regression in new compiler toolset being used to compile Roslyn?

I'll work on reverting it and then investigating.

jaredpar commented:
Yeah It goes away with `/p:UseRoslynAnalyzers=false`

mavasani commented:
@CyrusNajmabadi @jaredpar - this seems to a very recent perf regression in IOperation where we removed the IOperation.IsInvalid flag (which was using the BoundNode.HasErrors property) and replaced it with an extension method that calls GetDiagnostics on the node. Tagging @heejaechang, we need to find an alternate approach for deterimining IsInvalid. A trace of build with analyzers shows:

```
 Name                                              	Exc %	 Exc	Inc %	      Inc
analyzer.utilities!IOperationExtensions.IsInvalid 	  0.4	 304	 82.2	   64,641
```

Let me fix this quickly and see if moving to newer analyzer package resolves this issue.

sharwell commented:
:memo: @mavasani Within the GetDiagnostics call, the expensive operation was `ClsComplianceChecker`.
